linq4j/src/main/java/org/apache/calcite/linq4j/TopNHeap.java

```diff
@@ -0,0 +1,242 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.linq4j;
+
+import org.apache.calcite.linq4j.function.Function1;
+
+import java.util.Comparator;
+
+/**
+ * Implementation of a stable binary heap with a fetch and an offset.
+ * Stable means that if two items are considered equal,
+ * they will appear in the same order as they were offered to the heap.
+ *
+ * @param <TSource> type of the element that will be added to the heap
+ * @param <TKey> type of the key, which the comparator will use for comparisons
+ */
+public class TopNHeap<TSource, TKey> {
+  static final int MAX_INIT_ARRAY_SIZE = 1024;
+  static final int ROOT = 1;
+
+  private final Function1<TSource, TKey> keyFn;
+  private final Comparator<TKey> cmp;
+  private final int fetch;
+  private final int offset;
+  private final int maxSize;
+
+  TKey headKey = null;
+  int size = 0;
+  int time = -Integer.MIN_VALUE;
+
+  /** Heap with 1-based index, heap[0] is not used */
+  TSource[] heap;
+  /** Stores the order of arrival of the elements in the heap */
+  int[] order;
+
+  public TopNHeap(
+      Function1<TSource, TKey> keySelector,
+      Comparator<TKey> comparator,
+      int fetch,
+      int offset) {
+    this.keyFn = keySelector;
+    this.cmp = comparator;
+    long tmp = (long) fetch + offset;
+    this.maxSize = tmp > Integer.MAX_VALUE - ROOT ? Integer.MAX_VALUE - ROOT : (int) tmp;
+
+    //
+    int arrayLen = this.maxSize + ROOT;
+    if (arrayLen > 2 * MAX_INIT_ARRAY_SIZE) {
+      arrayLen = MAX_INIT_ARRAY_SIZE;
+    }
+
+    @SuppressWarnings("unchecked")
+    TSource[] t = (TSource[]) new Object[arrayLen];
+    this.heap = t;
+    this.order = new int[arrayLen];
+    this.fetch = fetch;
+    this.offset = offset;
+  }
+
+  /**
+   * Offers a new item to the heap.
+   */
+  public void offer(TSource o) {
+    // if size < maxSize
+    if (this.size < this.maxSize) {
+      this.size++;
+      if (this.size >= this.heap.length) {
+        this.growSize();
+      }
+
+      this.shiftUp(this.size, o, this.time++);
+      return;
+    }
+
+    // check head
+    TKey elKey = this.keyFn.apply(o);
+    int c = this.cmp.compare(elKey, this.headKey);
+    if (c >= 0) {
+      return;
+    }
+
+    this.shiftDown(ROOT, o, this.time++, elKey);
+  }
+
+  /**
+   * Returns an array with at most 'fetch' entries, that contains
+   * the sorted entries of the items added to the heap, while
+   * skipping the first 'offset' elements.
+   */
+  public Object[] getResult() {
+    int len = Math.min(this.size - this.offset, this.fetch);
+    len = Math.max(0, len);
+    @SuppressWarnings("unchecked")
+    TSource[] result = (TSource[]) new Object[len];
+    for (int i = result.length - 1; i >= 0; i--) {
+      result[i] = this.poll();
+    }
+    return result;
+  }
+
+  // --------------------------------------------------------------------------------
```

Comment:
minor: I think it would be better to add a meaningful comment here, e.g. `// ~ Private helper methods ------`